src/test/ui/type-alias-impl-trait/fallback.rs

```diff
@@ -0,0 +1,22 @@
+// Tests that we correctly handle the instantiated
+// inference variable being completely unconstrained.
+//
+// check-pass
```

Comment:
My understanding of the RFC (at least the spirit of it, it doesn't elaborate on cases like these, cc @cramertj) is that each function fully determines each referenced opaque type. In this case `_make_iter` does fully determine the opaque type `Foo`. However, `_produce`, without consulting other functions (imagine `_make_iter` did not exist), can only determine that `Wrapper::Second: Wrapper<?opaque_infer_var_0>`.

Other notable examples which we may consult are (the first of which you included as a snippet in your code):
```rust
fn alpha() -> Option<impl Copy> { None }
//~^ ERROR type annotations needed: cannot resolve `_: std::marker::Copy`
```
and:
```rust
fn main() {
    let beta = None;
    //~^ ERROR type annotations needed for `std::option::Option<T>`
}
```

In conclusion, while this is behavior that may be desirable, I don't think it follows from https://github.com/rust-lang/rfcs/pull/2071's notion of each function independently determining opaque types, and so this test shouldn't be `// check-pass`.

Reply:
@Centril: I believe the RFC explicitly alows this behavior:

 > A body or initializer must either fully constrain or place no constraints upon a given existential type.

See the `return_foo` example in the RFC.

Reply:
Hmm; that does seem to follow yes. Can you amend the test with the rationale given here as a comment?

Reply:
Updated